sql/core/src/main/scala/org/apache/spark/sql/execution/CacheManager.scala

```diff
@@ -97,13 +106,13 @@ private[sql] class CacheManager(sqlContext: SQLContext) extends Logging {
       logWarning("Asked to cache already cached data.")
     } else {
       cachedData +=
-        CachedData(
+        new CachedData(
           planToCache,
           InMemoryRelation(
             sqlContext.conf.useCompression,
             sqlContext.conf.columnBatchSize,
             storageLevel,
-            sqlContext.executePlan(query.logicalPlan).executedPlan,
+            sqlContext.executePlan(planToCache).executedPlan,
```

Comment:
This is just an optimization, and we don't want to re-analysis the logical plan as it's done right before the calls.

When you check the full code of the function `cacheQuery`, you will see the `planToCache` is just a local variable, not a logical plan cached by somewhere else, it should not be the case you were talking.